src/coredump/coredump.c

```diff
@@ -1115,18 +1115,43 @@ static int process_socket(int fd) {
                                 /* We have all FDs we need let's take a shortcut here. */
                                 break;
                         } else {
+                                struct cmsghdr *cmsg;
+                                unsigned n_fds = 0;
+
                                 found = cmsg_find(&mh, SOL_SOCKET, SCM_RIGHTS, CMSG_LEN(sizeof(int)));
-                                if (found)
+                                if (first && found) {
+                                        /* This is the first message that carries file descriptors, maybe there will be
+                                         * one more that actually contains array of two descriptors. */
+                                        assert(input_fd < 0);
+
                                         input_fd = *CMSG_TYPED_DATA(found, int);
-                        }
+                                        first = false;
+
+                                        continue;
+                                } else if (first && !found) {
+                                        /* This is the first message of zero length and it has no file descriptor,
+                                         * this is the protocol violation so let's bail out. */
+                                        cmsg_close_all(&mh);
+                                        r = log_error_errno(SYNTHETIC_ERRNO(EBADMSG),
+                                                    "Received zero length message with no file descriptor.");
+                                        goto finish;
+                                }
```

Comment:
both bracnhes here are conditioned on `first` being true? place in the same common if branch?

Reply:
Unaddressed. How about something like the following?
```c
if (first) {
        found = cmsg_find(...);
        if (!found) {
                cmsg_close_all(&mh);
                r = log_error_errno(...);
                goto finish;
        }

        assert(input_fd < 0);
        input_fd = ...;
        first = false;
        continue;
}
```